File: src/main/java/frc/robot/subsystems/swerve/module/ModuleIOHybridFXS.java
package frc.robot.subsystems.swerve.module;

import com.ctre.phoenix6.BaseStatusSignal;
import com.ctre.phoenix6.configs.CANcoderConfiguration;
import com.ctre.phoenix6.configs.MotorOutputConfigs;
import com.ctre.phoenix6.configs.Slot0Configs;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.configs.TalonFXSConfiguration;
import com.ctre.phoenix6.controls.MotionMagicVoltage;
import com.ctre.phoenix6.controls.TorqueCurrentFOC;
import com.ctre.phoenix6.controls.VelocityTorqueCurrentFOC;
import com.ctre.phoenix6.controls.VoltageOut;
import com.ctre.phoenix6.hardware.CANcoder;
import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.hardware.TalonFXS;
import com.ctre.phoenix6.signals.AdvancedHallSupportValue;
import com.ctre.phoenix6.signals.ExternalFeedbackSensorSourceValue;
import com.ctre.phoenix6.signals.InvertedValue;
import com.ctre.phoenix6.signals.MotorArrangementValue;
import com.ctre.phoenix6.signals.NeutralModeValue;
import com.ctre.phoenix6.StatusSignal;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.units.Units;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Distance;
import edu.wpi.first.units.measure.Voltage;
import lib.math.controllers.gains.FeedforwardGains;
import lib.math.controllers.gains.PIDGains;

/** Real robot module IO layer that uses Talon FX drive with Talon FXS steering. */
public final class ModuleIOHybridFXS implements ModuleIO {
  private final TalonFX driveMotor;
  private final TalonFXS turnMotor;
  private final CANcoder absoluteEncoder;
  private final double wheelRadiusMeters;

  private final VoltageOut openLoopDriveRequest = new VoltageOut(0.0);
  private final TorqueCurrentFOC openLoopTorqueRequest = new TorqueCurrentFOC(0.0);
  private final VelocityTorqueCurrentFOC closedLoopDriveRequest = new VelocityTorqueCurrentFOC(0.0);
  private final MotionMagicVoltage closedLoopTurnRequest = new MotionMagicVoltage(0.0);

  private final StatusSignal<Angle> drivePosition;
  private final StatusSignal<AngularVelocity> driveVelocity;
  private final StatusSignal<Voltage> driveSupplyVolts;
  private final StatusSignal<Voltage> driveMotorVolts;
  private final StatusSignal<Current> driveStatorCurrent;
  private final StatusSignal<Current> driveSupplyCurrent;
  private final StatusSignal<Current> driveTorqueCurrent;

  private final StatusSignal<Angle> turnPosition;
  private final StatusSignal<AngularVelocity> turnVelocity;
  private final StatusSignal<Voltage> turnAppliedVoltage;
  private final StatusSignal<Current> turnSupplyCurrent;
  private final StatusSignal<Current> turnStatorCurrent;

  private final StatusSignal<Angle> absoluteTurnPosition;

  public ModuleIOHybridFXS(
      int driveID,
      double driveGearing,
      boolean driveInverted,
      FeedforwardGains driveFF,
      PIDGains drivePID,
      int turnID,
      double turnGearing,
      boolean turnInverted,
      FeedforwardGains turnFF,
      PIDGains turnPID,
      int encoderID,
      Rotation2d encoderOffset,
      Distance wheelRadius) {
    this.wheelRadiusMeters = wheelRadius.in(Units.Meter);

    driveMotor = new TalonFX(driveID, "canivore");
    TalonFXConfiguration driveConfig = new TalonFXConfiguration();
    driveConfig.Feedback.SensorToMechanismRatio = driveGearing;
    driveConfig.Slot0.kP = drivePID.getKP();
    driveConfig.Slot0.kI = drivePID.getKI();
    driveConfig.Slot0.kD = drivePID.getKD();
    driveConfig.Slot0.kV = driveFF.getKV();
    driveConfig.Slot0.kA = driveFF.getKA();
    driveConfig.Slot0.kS = driveFF.getKS();
    driveConfig.CurrentLimits.StatorCurrentLimit = 120.0;
    driveConfig.CurrentLimits.SupplyCurrentLimit = 40.0;
    driveConfig.CurrentLimits.StatorCurrentLimitEnable = true;
    driveConfig.CurrentLimits.SupplyCurrentLimitEnable = true;
    driveConfig.TorqueCurrent.withPeakForwardTorqueCurrent(120.0);
    driveConfig.TorqueCurrent.withPeakReverseTorqueCurrent(-120.0);
    driveConfig.MotorOutput.Inverted =
        driveInverted ? InvertedValue.Clockwise_Positive : InvertedValue.CounterClockwise_Positive;
    driveConfig.MotorOutput.NeutralMode = NeutralModeValue.Brake;
    driveMotor.getConfigurator().apply(driveConfig);

    drivePosition = driveMotor.getPosition().clone();
    driveVelocity = driveMotor.getVelocity().clone();
    driveSupplyVolts = driveMotor.getSupplyVoltage().clone();
    driveMotorVolts = driveMotor.getMotorVoltage().clone();
    driveStatorCurrent = driveMotor.getStatorCurrent().clone();
    driveSupplyCurrent = driveMotor.getSupplyCurrent().clone();
    driveTorqueCurrent = driveMotor.getTorqueCurrent().clone();

    turnMotor = new TalonFXS(turnID, "canivore");
    TalonFXSConfiguration turnConfig = new TalonFXSConfiguration();
    turnConfig.ExternalFeedback.RotorToSensorRatio = turnGearing;
    turnConfig.ExternalFeedback.FeedbackRemoteSensorID = encoderID;
    turnConfig.ExternalFeedback.ExternalFeedbackSensorSource =
        ExternalFeedbackSensorSourceValue.FusedCANcoder;
    turnConfig.Commutation.MotorArrangement = MotorArrangementValue.NEO_JST;
    turnConfig.Commutation.AdvancedHallSupport = AdvancedHallSupportValue.Enabled;
    turnConfig.ClosedLoopGeneral.ContinuousWrap = true;
    turnConfig.MotorOutput.NeutralMode = NeutralModeValue.Brake;
    turnConfig.MotorOutput.Inverted =
        turnInverted ? InvertedValue.Clockwise_Positive : InvertedValue.CounterClockwise_Positive;
    turnConfig.Slot0.kP = turnPID.getKP();
    turnConfig.Slot0.kI = turnPID.getKI();
    turnConfig.Slot0.kD = turnPID.getKD();
    turnConfig.Slot0.kV = turnFF.getKV();
    turnConfig.Slot0.kA = turnFF.getKA();
    turnConfig.Slot0.kS = turnFF.getKS();
    turnConfig.MotionMagic.withMotionMagicCruiseVelocity(Units.RPM.of(260.0));
    turnConfig.MotionMagic.withMotionMagicAcceleration(Units.RotationsPerSecondPerSecond.of(500.0));
    turnMotor.getConfigurator().apply(turnConfig);

    turnPosition = turnMotor.getPosition().clone();
    turnVelocity = turnMotor.getVelocity().clone();
    turnAppliedVoltage = turnMotor.getMotorVoltage().clone();
    turnSupplyCurrent = turnMotor.getSupplyCurrent().clone();
    turnStatorCurrent = turnMotor.getStatorCurrent().clone();

    absoluteEncoder = new CANcoder(encoderID, "canivore");
    CANcoderConfiguration encoderConfig = new CANcoderConfiguration();
    encoderConfig.MagnetSensor.MagnetOffset = encoderOffset.getRotations();
    absoluteEncoder.getConfigurator().apply(encoderConfig);
    absoluteTurnPosition = absoluteEncoder.getAbsolutePosition().clone();
  }

  @Override
  public void updateInputs(ModuleIOInputs inputs) {
    inputs.driveMotorConnected =
        BaseStatusSignal.refreshAll(
                    drivePosition,
                    driveVelocity,
                    driveSupplyVolts,
                    driveMotorVolts,
                    driveStatorCurrent,
                    driveSupplyCurrent,
                    driveTorqueCurrent)
                .isOK();

    inputs.turnMotorConnected =
        BaseStatusSignal.refreshAll(
                    turnPosition,
                    turnVelocity,
                    turnSupplyCurrent,
                    turnAppliedVoltage,
                    turnStatorCurrent)
                .isOK();

    inputs.absoluteEncoderConnected =
        BaseStatusSignal.refreshAll(absoluteTurnPosition).isOK();

    inputs.driveVelocityRadPerSec = driveVelocity.getValueAsDouble();
    inputs.driveVelocityMetersPerSec = inputs.driveVelocityRadPerSec * wheelRadiusMeters;
    inputs.drivePositionRad = drivePosition.getValueAsDouble();
    inputs.drivePositionMeters = inputs.drivePositionRad * wheelRadiusMeters;
    inputs.driveAppliedVolts = driveMotorVolts.getValueAsDouble();
    inputs.driveStatorCurrentAmps = driveStatorCurrent.getValueAsDouble();
    inputs.driveSupplyCurrentAmps = driveSupplyCurrent.getValueAsDouble();
    inputs.driveTorqueCurrentAmps = driveTorqueCurrent.getValueAsDouble();

    inputs.turnPosition = Rotation2d.fromRotations(turnPosition.getValueAsDouble());
    inputs.absoluteTurnPosition =
        Rotation2d.fromRotations(absoluteTurnPosition.getValueAsDouble());
    inputs.turnVelocityRadPerSec = turnVelocity.getValueAsDouble();
    inputs.turnAppliedVolts = turnAppliedVoltage.getValueAsDouble();
    inputs.turnStatorCurrentAmps = turnStatorCurrent.getValueAsDouble();
    inputs.turnSupplyCurrentAmps = turnSupplyCurrent.getValueAsDouble();
  }

  @Override
  public void setDriveVoltage(double volts) {
    driveMotor.setControl(openLoopDriveRequest.withOutput(volts));
  }

  @Override
  public void setTurnVoltage(double volts) {
    turnMotor.setVoltage(volts);
  }

  @Override
  public void setTurnPosition(Rotation2d position) {
    turnMotor.setControl(closedLoopTurnRequest.withPosition(position.getRotations()));
  }

  @Override
  public void setDriveVelocity(double velocityMetersPerSec) {
    setDriveVelocity(velocityMetersPerSec, 0.0);
  }

  @Override
  public void setDriveVelocity(double velocityMetersPerSec, double torqueCurrentAmps) {
    double velocityRadPerSec = velocityMetersPerSec / wheelRadiusMeters;
    AngularVelocity velocityMeasure = Units.RadiansPerSecond.of(velocityRadPerSec);
    Current ff = Units.Amps.of(torqueCurrentAmps);
    driveMotor.setControl(closedLoopDriveRequest.withVelocity(velocityMeasure).withFeedForward(ff));
  }

  @Override
  public void reset() {
    driveMotor.setPosition(0.0);
  }

  @Override
  public void stop() {
    driveMotor.stopMotor();
    turnMotor.stopMotor();
  }

  @Override
  public void setDriveBrake(boolean enabled) {
    TalonFXConfiguration config = new TalonFXConfiguration();
    driveMotor.getConfigurator().refresh(config);
    config.MotorOutput.NeutralMode = enabled ? NeutralModeValue.Brake : NeutralModeValue.Coast;
    driveMotor.getConfigurator().apply(config);
  }

  @Override
  public void setTurnBrake(boolean enabled) {
    MotorOutputConfigs config = new MotorOutputConfigs();
    turnMotor.getConfigurator().refresh(config);
    turnMotor
        .getConfigurator()
        .apply(config.withNeutralMode(enabled ? NeutralModeValue.Brake : NeutralModeValue.Coast));
  }

  @Override
  public void setDrivePID(PIDGains gains) {
    Slot0Configs config = new Slot0Configs();
    driveMotor.getConfigurator().refresh(config);
    driveMotor
        .getConfigurator()
        .apply(config.withKP(gains.getKP()).withKI(gains.getKI()).withKD(gains.getKD()));
  }

  @Override
  public void setSteerPID(PIDGains gains) {
    Slot0Configs config = new Slot0Configs();
    turnMotor.getConfigurator().refresh(config);
    turnMotor
        .getConfigurator()
        .apply(config.withKP(gains.getKP()).withKI(gains.getKI()).withKD(gains.getKD()));
  }

  @Override
  public void setDriveCurrent(double currentAmps) {
    driveMotor.setControl(openLoopTorqueRequest.withOutput(Units.Amps.of(currentAmps)));
  }
}

File: src/main/java/frc/robot/subsystems/swerve/module/ModuleIOHybridFXS.kt
package frc.robot.subsystems.swerve.module

import com.ctre.phoenix6.BaseStatusSignal
import com.ctre.phoenix6.configs.CANcoderConfiguration
import com.ctre.phoenix6.configs.MotorOutputConfigs
import com.ctre.phoenix6.configs.Slot0Configs
import com.ctre.phoenix6.configs.TalonFXConfiguration
import com.ctre.phoenix6.configs.TalonFXSConfiguration
import com.ctre.phoenix6.controls.MotionMagicExpoVoltage
import com.ctre.phoenix6.controls.TorqueCurrentFOC
import com.ctre.phoenix6.controls.VelocityTorqueCurrentFOC
import com.ctre.phoenix6.controls.VoltageOut
import com.ctre.phoenix6.hardware.CANcoder
import com.ctre.phoenix6.hardware.TalonFX
import com.ctre.phoenix6.hardware.TalonFXS
import com.ctre.phoenix6.signals.*
import com.revrobotics.spark.ClosedLoopSlot
import com.revrobotics.spark.SparkBase
import com.revrobotics.spark.config.SparkBaseConfig
import com.revrobotics.spark.config.SparkMaxConfig
import edu.wpi.first.math.geometry.Rotation2d
import edu.wpi.first.units.Units.*
import edu.wpi.first.units.measure.Current
import edu.wpi.first.units.measure.Distance
import edu.wpi.first.units.measure.LinearVelocity
import edu.wpi.first.units.measure.Voltage
import lib.controllers.gains.FeedforwardGains
import lib.controllers.gains.PIDGains
import lib.math.units.into
import lib.math.units.measuredIn

class ModuleIOHybridFXS(
    val driveID: Int,
    val driveGearing: Double,
    val driveInverted: Boolean,
    val driveFF: FeedforwardGains,
    val drivePID: PIDGains,

    val turnID: Int,
    val turnGearing: Double,
    val turnInverted: Boolean,
    val turnFF: FeedforwardGains,
    val turnPID: PIDGains,

    val encoderID: Int,
    val encoderOffset: Rotation2d,

    val wheelRadius: Distance
) : ModuleIO {
    private val driveMotor: TalonFX = TalonFX(driveID).apply {
        val config: TalonFXConfiguration = TalonFXConfiguration()

        config.Feedback.SensorToMechanismRatio = driveGearing

        config.Slot0.kP = drivePID.kP
        config.Slot0.kI = drivePID.kI
        config.Slot0.kD = drivePID.kD
        config.Slot0.kV = driveFF.kV
        config.Slot0.kA = driveFF.kA
        config.Slot0.kS = driveFF.kS

        config.CurrentLimits.StatorCurrentLimit = 120.0
        config.CurrentLimits.SupplyCurrentLimit = 40.0
        config.CurrentLimits.StatorCurrentLimitEnable = true
        config.CurrentLimits.SupplyCurrentLimitEnable = true
        config.TorqueCurrent.withPeakForwardTorqueCurrent(120.0)
        config.TorqueCurrent.withPeakReverseTorqueCurrent(-120.0)

        config.MotorOutput.Inverted =
            if (driveInverted) InvertedValue.CounterClockwise_Positive
            else InvertedValue.Clockwise_Positive

        config.MotorOutput.NeutralMode = NeutralModeValue.Brake

        configurator.apply(config)
    }

    private val drivePosition = driveMotor.position.clone()
    private val driveVelocity = driveMotor.velocity.clone()
    private val driveSupplyVolts = driveMotor.supplyVoltage.clone()
    private val driveMotorVolts = driveMotor.motorVoltage.clone()
    private val driveStatorCurrent = driveMotor.statorCurrent.clone()
    private val driveSupplyCurrent = driveMotor.supplyCurrent.clone()
    private val driveTorqueCurrent = driveMotor.torqueCurrent.clone()

    private val turnMotor = TalonFXS(turnID).apply {
        val config: TalonFXSConfiguration = TalonFXSConfiguration()

        config.ExternalFeedback.RotorToSensorRatio = turnGearing
        config.ExternalFeedback.FeedbackRemoteSensorID = encoderID
        config.ExternalFeedback.ExternalFeedbackSensorSource = ExternalFeedbackSensorSourceValue.FusedCANcoder

        config.Commutation.MotorArrangement = MotorArrangementValue.NEO_JST
        config.Commutation.AdvancedHallSupport = AdvancedHallSupportValue.Enabled

        config.ClosedLoopGeneral.ContinuousWrap = true
        config.MotorOutput.NeutralMode = NeutralModeValue.Brake

        config.Slot0.kP = drivePID.kP
        config.Slot0.kI = drivePID.kI
        config.Slot0.kD = drivePID.kD
        config.Slot0.kV = driveFF.kV
        config.Slot0.kA = driveFF.kA
        config.Slot0.kS = driveFF.kS

        configurator.apply(config)
    }

    private val turnPosition = turnMotor.position.clone()
    private val turnVelocity = turnMotor.velocity.clone()
    private val turnAppliedVoltage = turnMotor.motorVoltage.clone()
    private val turnSupplyCurrent = turnMotor.supplyCurrent.clone()
    private val turnStatorCurrent = turnMotor.statorCurrent.clone()

    private val absEncoder: CANcoder = CANcoder(encoderID).apply {
        val config = CANcoderConfiguration()

        config.MagnetSensor.MagnetOffset = encoderOffset.rotations


        configurator.apply(config)
    }

    val absoluteTurnPosition = absEncoder.absolutePosition.clone()

    val openLoopDriveRequest: VoltageOut = VoltageOut(0.0)
    val openLoopTorqueRequest: TorqueCurrentFOC = TorqueCurrentFOC(0.0)
    val closedLoopDriveRequest: VelocityTorqueCurrentFOC = VelocityTorqueCurrentFOC(0.0)

    val closedLoopTurnRequest: MotionMagicExpoVoltage = MotionMagicExpoVoltage(0.0)

    override fun updateInputs(inputs: ModuleIO.ModuleInputs) {
        inputs.isDriveMotorConnected = BaseStatusSignal.refreshAll(
            drivePosition,
            driveVelocity,
            driveSupplyVolts,
            driveMotorVolts,
            driveStatorCurrent,
            driveSupplyCurrent,
            driveTorqueCurrent
        ).isOK

        inputs.isTurnMotorConnected = BaseStatusSignal.refreshAll(
            turnPosition,
            turnVelocity,
            turnSupplyCurrent,
            turnAppliedVoltage,
            turnStatorCurrent
        ).isOK

        inputs.isAbsoluteEncoderConnected = BaseStatusSignal.refreshAll(
            absoluteTurnPosition
        ).isOK

        inputs.driveVelocity.mut_replace(
            (driveVelocity.value into RadiansPerSecond) * (wheelRadius into Meters),
            MetersPerSecond
        )

        inputs.drivePosition.mut_replace(
            (drivePosition.value into Radians) * (wheelRadius into Meters),
            Meters
        )

        inputs.drivePositionAngular.mut_replace(drivePosition.value)

        inputs.driveAppliedVoltage.mut_replace(
            driveMotorVolts.value
        )

        inputs.driveStatorCurrent.mut_replace(
            driveStatorCurrent.value
        )

        inputs.driveSupplyCurrent.mut_replace(
            driveSupplyCurrent.value
        )

        inputs.driveTorqueCurrent.mut_replace(
            driveTorqueCurrent.value
        )

        inputs.turnPosition = Rotation2d.fromRotations(turnPosition.valueAsDouble)

        inputs.absoluteTurnPosition = Rotation2d.fromRotations(absoluteTurnPosition.valueAsDouble)

        inputs.turnVelocity.mut_replace(turnVelocity.value)

        inputs.turnAppliedVoltage.mut_replace(turnAppliedVoltage.value)

        inputs.turnStatorCurrent.mut_replace(turnStatorCurrent.value)
        inputs.turnSupplyCurrent.mut_replace(turnSupplyCurrent.value)
    }

    override fun setDriveVoltage(volts: Voltage) {
        driveMotor.setControl(openLoopDriveRequest.withOutput(volts))
    }

    override fun setTurnVoltage(volts: Voltage) {
        turnMotor.setVoltage(volts into Volts)
    }

    override fun setTurnPosition(position: Rotation2d) {
        turnMotor.setControl(closedLoopTurnRequest.withPosition(position.rotations))
    }

    override fun setDriveVelocity(velocity: LinearVelocity) {
        setDriveVelocity(velocity, Amps.zero())
    }

    /**
     * Send a velocity setpoint to the drive motor
     *
     * @param velocity The velocity to set the motor to
     * @param torqueCurrentFF The feedforward torque current to apply
     */
    override fun setDriveVelocity(velocity: LinearVelocity, torqueCurrentFF: Current) {
        driveMotor.setControl(
            closedLoopDriveRequest.withVelocity(
                (velocity into MetersPerSecond) / (wheelRadius into Meters) measuredIn RadiansPerSecond,
            ).withFeedForward(torqueCurrentFF),
        )
    }

    override fun reset() {
        driveMotor.setPosition(0.0)
    }

    override fun stop() {
        driveMotor.stopMotor()
        turnMotor.stopMotor()
    }

    override fun setDriveBrake(enabled: Boolean) {
        val config = TalonFXConfiguration()
        driveMotor.configurator.refresh(config)

        config.MotorOutput.NeutralMode = if (enabled) NeutralModeValue.Brake else NeutralModeValue.Coast
    }

    override fun setTurnBrake(enabled: Boolean) {
        val config = MotorOutputConfigs()
        turnMotor.configurator.refresh(config)
        turnMotor.configurator.apply(config.withNeutralMode(if(enabled) NeutralModeValue.Brake else NeutralModeValue.Coast))
    }

    override fun setDrivePID(gains: PIDGains) {
        val config = Slot0Configs()
        driveMotor.configurator.refresh(config)
        driveMotor
            .configurator
            .apply(
                config
                    .withKP(gains.kP)
                    .withKI(gains.kI)
                    .withKD(gains.kD)
            )
    }

    override fun setSteerPID(gains: PIDGains) {
        val config = Slot0Configs()
        turnMotor.configurator.refresh(config)
        turnMotor
            .configurator
            .apply(
                config
                    .withKP(gains.kP)
                    .withKI(gains.kI)
                    .withKD(gains.kD))
    }

    override fun setDriveCurrent(current: Current) {
        driveMotor.setControl(openLoopTorqueRequest.withOutput(current))
    }
}